File: mcp-code-intelligence-python/src/mcp_code_intel/orchestration/cache/adaptive_cache.py
"""AdaptiveTokenCache — self-learning fuzzy token cache for find_tools."""

from __future__ import annotations

import sys
from pathlib import Path

from .cache_entry import CacheEntry
from .invalidation import compute_token_overlap, evict_lru, invalidate_stale
from .persistence import DebouncedPersistence

DEFAULT_THRESHOLD = 0.80
DEFAULT_CONFIDENCE = 0.75


class AdaptiveTokenCache:
    """In-memory cache with fuzzy token matching and debounced persistence."""

    def __init__(self, cache_path: Path, debounce_s: float = 5.0) -> None:
        self._persistence = DebouncedPersistence(cache_path, debounce_s)
        self._entries: list[CacheEntry] = []
        self._loaded = False
        self._hits = 0
        self._misses = 0

    def find_fuzzy(
        self, tokens: set[str], threshold: float = DEFAULT_THRESHOLD
    ) -> CacheEntry | None:
        """Find best cache entry with ≥threshold token overlap."""
        self._ensure_loaded()
        best: CacheEntry | None = None
        best_overlap = 0.0
        for entry in self._entries:
            overlap = compute_token_overlap(tokens, entry.tokens)
            if overlap >= threshold and overlap > best_overlap:
                best = entry
                best_overlap = overlap
        if best is not None:
            self._hits += 1
            best.touch()
        else:
            self._misses += 1
        return best

    def add(
        self, tokens: set[str], tool_name: str, score: float, registry_hash: str
    ) -> None:
        """Add or update cache entry from embedding result."""
        self._ensure_loaded()
        if score < DEFAULT_CONFIDENCE:
            return
        existing = self._find_exact(tool_name, tokens)
        if existing:
            self._merge_entry(existing, tokens, score)
        else:
            entry = CacheEntry(
                tokens=frozenset(tokens),
                tool_name=tool_name,
                score=score,
                registry_hash=registry_hash,
            )
            self._entries.append(entry)
        self._entries = evict_lru(self._entries)

    def invalidate_stale(self, current_hash: str) -> int:
        """Remove entries with mismatched registry hash."""
        self._ensure_loaded()
        self._entries, removed = invalidate_stale(self._entries, current_hash)
        if removed > 0:
            self.schedule_persist()
        return removed

    def schedule_persist(self) -> None:
        """Schedule debounced write to disk."""
        data = self._serialize()
        self._persistence.schedule_write(data)

    def load(self) -> None:
        """Force load from disk."""
        self._do_load()

    @property
    def size(self) -> int:
        """Current number of active entries."""
        self._ensure_loaded()
        return len(self._entries)

    @property
    def hit_rate(self) -> float:
        """Cache hit rate (0.0 to 1.0)."""
        total = self._hits + self._misses
        return self._hits / total if total > 0 else 0.0

    def _ensure_loaded(self) -> None:
        """Lazy load on first access."""
        if not self._loaded:
            self._do_load()

    def _do_load(self) -> None:
        """Load cache entries from disk."""
        self._loaded = True
        data = self._persistence.load()
        if data is None:
            return
        raw_entries = data.get("entries", [])
        self._entries = [CacheEntry.from_dict(e) for e in raw_entries]
        _log(f"Loaded {len(self._entries)} cache entries")

    def _find_exact(self, tool_name: str, tokens: set[str]) -> CacheEntry | None:
        """Find existing entry for same tool with high overlap."""
        for entry in self._entries:
            if entry.tool_name == tool_name:
                overlap = compute_token_overlap(tokens, entry.tokens)
                if overlap >= 0.6:
                    return entry
        return None

    def _merge_entry(self, entry: CacheEntry, new_tokens: set[str], score: float) -> None:
        """Merge new tokens into existing entry, keep highest score."""
        merged = entry.tokens | frozenset(new_tokens)
        entry.tokens = merged
        entry.score = max(entry.score, score)
        entry.touch()

    def _serialize(self) -> dict:
        """Serialize cache to JSON-compatible dict."""
        return {
            "version": 1,
            "entries": [e.to_dict() for e in self._entries],
        }


def _log(msg: str) -> None:
    print(f"[adaptive-cache] {msg}", file=sys.stderr, flush=True)

File: backend/mcp-code-intelligence-nodejs/src/orchestration/cache/adaptive-cache.ts
/**
 * AdaptiveTokenCache — self-learning fuzzy token cache for find_tools.
 */

import { CacheEntry, createCacheEntry, touchEntry, entryToJson, entryFromJson } from './cache-entry.js';
import { computeTokenOverlap, evictLru, invalidateStale } from './invalidation.js';
import { DebouncedPersistence } from './persistence.js';

const DEFAULT_THRESHOLD = 0.80;
const DEFAULT_CONFIDENCE = 0.75;

export class AdaptiveTokenCache {
  private persistence: DebouncedPersistence;
  private entries: CacheEntry[] = [];
  private loaded = false;
  private hits = 0;
  private misses = 0;

  constructor(cachePath: string, debounceS: number = 5.0) {
    this.persistence = new DebouncedPersistence(cachePath, debounceS);
  }

  /** Find best cache entry with ≥threshold token overlap. */
  findFuzzy(tokens: Set<string>, threshold: number = DEFAULT_THRESHOLD): CacheEntry | null {
    this.ensureLoaded();
    let best: CacheEntry | null = null;
    let bestOverlap = 0;
    for (const entry of this.entries) {
      const overlap = computeTokenOverlap(tokens, entry.tokens);
      if (overlap >= threshold && overlap > bestOverlap) {
        best = entry;
        bestOverlap = overlap;
      }
    }
    if (best) { this.hits++; touchEntry(best); }
    else { this.misses++; }
    return best;
  }

  /** Add or update cache entry from embedding result. */
  add(tokens: Set<string>, toolName: string, score: number, registryHash: string): void {
    this.ensureLoaded();
    if (score < DEFAULT_CONFIDENCE) return;
    const existing = this.findExact(toolName, tokens);
    if (existing) {
      this.mergeEntry(existing, tokens, score);
    } else {
      this.entries.push(createCacheEntry(tokens, toolName, score, registryHash));
    }
    this.entries = evictLru(this.entries);
  }

  /** Remove entries with mismatched registry hash. */
  invalidateStale(currentHash: string): number {
    this.ensureLoaded();
    const [kept, removed] = invalidateStale(this.entries, currentHash);
    this.entries = kept;
    if (removed > 0) this.schedulePersist();
    return removed;
  }

  /** Schedule debounced write to disk. */
  schedulePersist(): void {
    this.persistence.scheduleWrite(this.serialize());
  }

  /** Force load from disk. */
  load(): void { this.doLoad(); }

  get size(): number { this.ensureLoaded(); return this.entries.length; }
  get hitRate(): number {
    const total = this.hits + this.misses;
    return total > 0 ? this.hits / total : 0;
  }

  private ensureLoaded(): void { if (!this.loaded) this.doLoad(); }

  private doLoad(): void {
    this.loaded = true;
    const data = this.persistence.load();
    if (!data) return;
    const raw = data.entries ?? [];
    this.entries = raw.map((e: any) => entryFromJson(e));
    console.error(`[adaptive-cache] Loaded ${this.entries.length} cache entries`);
  }

  private findExact(toolName: string, tokens: Set<string>): CacheEntry | null {
    for (const entry of this.entries) {
      if (entry.toolName === toolName && computeTokenOverlap(tokens, entry.tokens) >= 0.6) {
        return entry;
      }
    }
    return null;
  }

  private mergeEntry(entry: CacheEntry, newTokens: Set<string>, score: number): void {
    for (const t of newTokens) entry.tokens.add(t);
    entry.score = Math.max(entry.score, score);
    touchEntry(entry);
  }

  private serialize(): Record<string, any> {
    return { version: 1, entries: this.entries.map(entryToJson) };
  }
}
